libcob/profiling.c

```diff
@@ -0,0 +1,464 @@
+
+	while (current_idx >= 0) {
+		int curr_proc = called_procedures[current_idx];
+		struct cob_prof_module *curr_info = called_runtimes[current_idx];
+
+		curr_info->procedure_recursions[curr_proc]--;
+		if (curr_info->procedure_recursions[curr_proc]==0){
+			curr_info->total_times[curr_proc] += t - start_times[current_idx];
+		}
+		current_idx--;
+		if (curr_proc == proc_idx && curr_info == info) {
+			return;
+		}
+	}
+}
+
+void
+cob_prof_enter_paragraph (struct cob_prof_module *info, int proc_idx)
+{
+	cob_prof_enter_procedure (info, proc_idx);
+}
+
+void
+cob_prof_enter_section (struct cob_prof_module *info, int proc_idx)
+{
+	if (!is_active || info == NULL) return;
+	/* We do not measure time on section enter/exit, we use the cumulative time
+	   of all paragraphs of the section */
+	info->called_count[proc_idx] ++;
+}
+
+void
+cob_prof_enter_program (struct cob_prof_module *info){
+	/* program always has index 0 */
+	cob_prof_enter_procedure (info, 0);
+}
+
+void
+cob_prof_exit_program (struct cob_prof_module *info){
+	/* program always has index 0 */
+	cob_prof_exit_procedure (info, 0);
+}
+
+void
+cob_prof_use_paragraph_entry (struct cob_prof_module *info,
+			      int paragraph_idx, int entry_idx){
+	if (!is_active || info == NULL) return;
+	info->called_count[entry_idx] ++;
+	cob_prof_enter_paragraph (info, paragraph_idx);
+}
+
+void
+cob_prof_exit_paragraph (struct cob_prof_module *info, int proc_idx)
+{
+	cob_prof_exit_procedure (info, proc_idx);
+}
+
+void
+cob_prof_exit_section (struct cob_prof_module *info, int proc_idx)
+{
+	/* For now, nothing to do */
+}
+
+void
+cob_prof_goto (struct cob_prof_module *info)
+{
+	int curr_proc;
+	struct cob_prof_module *curr_info;
+
+	if (!is_active || info == NULL) return;
+
+	curr_proc = called_procedures[current_idx];
+	curr_info = called_runtimes[current_idx];
+
+	if (curr_info->procedures[curr_proc].kind == COB_PROF_PROCEDURE_PARAGRAPH){
+		cob_prof_exit_procedure (curr_info, curr_proc);
+	}
+}
+
+static void
+print_monotonic_time (FILE *file, cob_ns_time t) {
+
+	cob_ns_time nanoseconds = t ;
+	unsigned int hours = 0;
+	unsigned int minutes = 0;
+	unsigned int seconds = 0;
+	unsigned int milliseconds = 0;
+
+	if (nanoseconds >= 1000000) {
+		milliseconds = t / 1000000;
+		nanoseconds %= 1000000;
+	}
+	if (milliseconds >= 1000) {
+		seconds = milliseconds / 1000;
+		milliseconds %= 1000;
+	}
+	if (seconds >= 60) {
+		minutes = seconds / 60;
+		seconds %= 60;
+	}
+	if (minutes >= 60) {
+		hours = minutes / 60;
+		minutes %= 60;
+	}
+
+	fprintf (file, "%Ld,", t);
+
+	if (hours > 0) {
+		fprintf (file, "%u h %u m %u s", hours, minutes, seconds);
+	} else if (minutes > 0) {
+		fprintf (file, "%u m %u s %u ms", minutes, seconds, milliseconds);
+	} else if (seconds > 0) {
+		fprintf (file, "%u s %u ms %Lu ns", seconds, milliseconds, nanoseconds);
+	} else if (milliseconds > 0) {
+		fprintf (file, "%u ms %Lu ns", milliseconds, nanoseconds);
+	} else {
+		fprintf (file, "%Lu ns", nanoseconds);
+	}
```

Comment:
Should this be specific to `LC_TIME`?

Reply:
I would assume `LC_TIME` is used for calendar times, not for differences of times.

Reply:
right, this would also mean to always print via strftime which has leading zeroes; the point of my question was "should we care", for example in both US/UK and Spain it seems the expected output would be without the separator space after the digit (only before the unit), while in German and French there is an expected space both before and after the unit.

But I'm ok with just using the English variant and don't care about locale settings here.

Just wondering: Maybe "human readable" would be using fractions for anything more grained than a second?
```diff
-5 m 46 s 915 ms
+5m 46.915s
-5 s 465 ms 915 ns
+5.465915s
-915 ns
+0.000915s
```
Not sure if this would be considered "more human readable - I'm leaving the decision "output of fractions" or "output more than one/two units" to you, but in any case please drop the space before the unit.

Reply:
To be honest, I am not even sure it makes sense to have a precision over 1/100. I would actually prefer to only print everything in seconds (to avoid the fact that minutes and hours are not factors of 10...)